src/backend/distributed/transaction/transaction_recovery.c

```diff
@@ -257,6 +260,44 @@ RecoverWorkerTransactions(WorkerNode *workerNode)
 			continue;
 		}
 
+		/* Check if the transaction is created by an outer transaction from another database */
+		bool outerXidIsNull = false;
+		Datum outerXidDatum = heap_getattr(heapTuple,
+										   Anum_pg_dist_transaction_outerxid,
+									  	   tupleDescriptor, &outerXidIsNull);
+		
+		if (!outerXidIsNull)
+		{
+			FullTransactionId outerFullXid = DatumGetFullTransactionId(outerXidDatum);
+			TransactionId outerXid = XidFromFullTransactionId(outerFullXid);
+			bool outerXactIsInProgress = TransactionIdIsInProgress(outerXid);
+			bool outerXactDidCommit = TransactionIdDidCommit(outerXid);
+			if (outerXactIsInProgress && !outerXactDidCommit)
+			{
+				/*
+				 * The transaction is initiated from an outer transaction and the outer
+				 * transaction is not yet committed, so we should not commit either
+				 */
+				 continue;
+			}
+			else if (!outerXactIsInProgress && !outerXactDidCommit)
+			{
+				/*
+				 * Since outer transaction isn't in progress and did not commit we need to
+				 * abort the prepared transaction too.
+				 */
+				hash_search(pendingTransactionSet, transactionName, HASH_REMOVE,
+							&foundPreparedTransactionBeforeCommit);
+			}
```

Comment:
I believe this logic is incorrect. As far as I can tell, the code in the `if (!recoveryFailed)` block below would now abort any transaction that matches the first case, because it aborts any transactions that are in the pendinTransactionSet that are not themselves in progress. And it would never actually run its abort code for any transaction that matches the second case, because the transaction is removed from the pendingTransactionSet.

One way I can think of to handle the first case correctly is to again check if the outer transaction is still in progress in the abort loop in the `if (!recoveryFailed)` block below. 

For the second case I think it would work to simply replace the hash removal wit a `continue;`.

Also its really non-obvious how this code works. So these cases should have better comments than they have now. Also failure tests are sorely needed, because maybe my suggestion is also incorrect. And the only way to know that is by having tests that execute these failure cases.